### PR TITLE
Add style transition API

### PR DESCRIPTION
**Pull request by donadigo**
Adds a new method to the DBus API called `transition_from_snapshot` for apps (and primarily the switchboard plug) to be able to create a smooth stylesheet transition. The transition will not happen when animations are disabled but the snapshot will still be created and added (that is it will stay for 200ms and hide without any animations). 

The code should be quite optimized for the task, I previously used a simple `Cogl.Texture.get_data` for creating a static texture but it turned out to be quite long to execute: transferring pixel data from GPU to CPU took ~15ms on my machine which is little bit above 60FPS but there were longer runs that caused visible lag when creating textures for maximized windows. We now copy the texture directly on the GPU which takes about ~2ms.

**Update**: the code has been updated to Cogl2 and to automatically transition when the style changes, however there still seems to be the problem of some elements flickering while transitioning. I think this is a matter of timing things correctly.

### CODE REVIEW

**Comment by danrabbit** [comment]
I guess an obvious question is, what about Wayland? Do we get window IDs like this once we move away from X?

**Comment by donadigo** [comment]
1. No one knows. See https://github.com/elementary/gala/issues/68. I'm not sure if mutter will do anything about this but it seems like there's no standard for those things either. I also don't know how mutter on wayland handles window ID's.

2. We can optionally separate the other part of the method to a new one but it will be still public though.

**Comment by donadigo** [comment]
@danrabbit updated.

**Comment by cassidyjames** [comment]
Does this need to be an external API at all? Can't Gala just listen to the setting and do this transition automatically?

**Comment by donadigo** [comment]
@cassidyjames yeah that and if gala listened to that setting you have to account for the fact that the signal could fire later than the theme change, causing a jarring flick before the actual transition. That's why it's safer to have the app call before actually changing the setting. If we had a Gtk fork / patch then perhaps that would be possible to do automatically.

**Comment by donadigo** [comment]
Branch updated to use the newest Cogl2 API and updated to master.

**Comment by donadigo** [comment]
Ready for review again.

**Comment by danrabbit** [comment]
Hm, I can't really see that this smooths anything out for me. Is there something I need to do other than install this branch and restart my session?

**Comment by donadigo** [comment]
@danrabbit hm, that's weird. Installing the branch & restarting should be enough. Is the transition like instant, or do you see at least some trace of the animation?

File: lib/Constants.vala [hunk 1]
```diff
@@ -30,5 +30,7 @@ namespace Gala {
         SNAP = 250,
         // Duration of the workspace switch animation
         WORKSPACE_SWITCH = 300,
+        // Style switch animation
+        STYLE_SWITCH = 200
```
Comment (cassidyjames) [inline lib/Constants.vala:34]: 200 would feel nice and snappy, but it seems to not be slow enough to mask all the GTK transitions. Something like 500 seems to mask them better.

```suggestion
        STYLE_SWITCH = 500
```

File: src/DBus.vala [hunk 2]
```diff
@@ -256,5 +296,125 @@ namespace Gala {
...
+            }
+        }
+
+        static async void transition_window (Meta.Window window, bool animate) {
+            /**
+             * Because Gtk takes time to apply the styling to all elements,
+             * we will simply define an arbitrary wait time for this to happen
+             * to smoothen out the transition.
+             */
+            const uint STYLE_CHANGE_COMPLETION_DURATION = 150;
```
Comment (cassidyjames) [inline src/DBus.vala:359]: I've been playing with this, and 300 seems to be long enough to mask WebKit transitioning as well which is a really nice effect.

```suggestion
            const uint STYLE_CHANGE_COMPLETION_DURATION = 300;
```

File: lib/Constants.vala (hunk 1 above)
Comment (cassidyjames) [reply lib/Constants.vala:34]: Oh, nevermind, ignore me. 200 works fine if we increase the delay below slightly.

**Comment by cassidyjames** [comment]
@donadigo would it be possible to resolve conflicts here? I'm not sure why we weren't seeing it work in the past, but now that we have all the other dark style bits merged in, this would be an incredibly nice effect.

**Review by JoseExposito — CHANGES_REQUESTED** [review]
I didn't notice there was a PR open with this functionality, definitely something really cool to have! 
It's most likely something related to my set up, but I'm seeing some noise in the panel and shadows:

https://user-images.githubusercontent.com/1335948/126907909-8545c1cd-9f28-4c14-886c-61a7b432f57d.mp4

**Comment by danirabbit** [comment]
Gonna close this since there was a lot more progress on #1242

**Pull request closed without merging**